Source/Core/DolphinQt/RenderWidget.cpp

```diff
@@ -487,38 +487,40 @@ void RenderWidget::PassEventToImGui(const QEvent* event)
     const QKeyEvent* key_event = static_cast<const QKeyEvent*>(event);
     const bool is_down = event->type() == QEvent::KeyPress;
     const u32 key = static_cast<u32>(key_event->key() & 0x1FF);
-    auto lock = g_renderer->GetImGuiLock();
-    if (key < std::size(ImGui::GetIO().KeysDown))
-      ImGui::GetIO().KeysDown[key] = is_down;
+
+    const char* chars = nullptr;
 
     if (is_down)
     {
       auto utf8 = key_event->text().toUtf8();
-      ImGui::GetIO().AddInputCharactersUTF8(utf8.constData());
+
+      if (utf8.size())
+        chars = utf8.constData();
     }
+
+    // Pass the key onto ImGui
```

Comment:
The problem is that this code is a bridge between Qt and ImGui. Presenter really has nothing to do with it.

However, because I was planning to move presenter (and therefor ImGui) to a new thread, I thought it would be wise to be able to gate these keys/mouse changes and only apply them once per frame.

Maybe I should beef up this abstraction. It bugged me that RenderWidget still ended up including <imgui.h> (just so it could get the constants for keys, for configuring the keymap).

I might create an intermediary API owned by presenter and convert things twice.

Reply:
Done. I've created a more solid abstraction (though it's still very imgui)